Adds API to GEP-91: Client Certificate Validation (#2273)
kubernetes-sigs/gateway-api: geps/gep-91.md

Reviewer: > This GEP proposes a way to validate the TLS certificate presented by the downstream client to the server
> (Gateway Listener in this case) during a [TLS Handshake Protocol][], also commonly referred to as mutual TLS (mTLS).

Could you elaborate?  I'm not sure everyone would conclude that this is commonly referred to as mutual TLS.

Reviewer: If you could also elaborate on what you mean by client and server in this proposal, that would be helpful.

Reviewer 2: @candita can you clarify why you think the client and server here needs elaboration. The server is a Gateway listener (it says that) and the client is then whoever is making requests to the Gateway. Is there something more subtle that you're looking for?

Reviewer 3: I think the issue here is that, while this is definitely a mutual handshake, calling something "mutual TLS" _can_ be taken to mean a very specific thing (dynamically-issued, frequently-rotated certificates mutually authenticated between services as a way to handle service-to-service authentication). This is definitely not that thing, but one thing that @candita and I have learned through _much longer than we would have liked_ experience is that you _really_ shouldn't use "mutual TLS" without some explanation of what sense you mean it in, or people will believe it's the wrong one.

Yes, I _know_ that "mutual TLS" in this case is technically correct, but it also can be misleading if, for example, a person is used to reading that phrase in a mesh context, not in a reverse-proxy-or-Ingress context.

Author: removed the controversial mTLS term from the descriptions

Reviewer: I would also remove the ambiguous phrase "downstream client", and replace it with something that clearly indicates what resource/s are connecting.

Author: I dont think downstream client is ambiguous, can you share any other alternatives ?

Reviewer 4: The term 'mTLS' is far older than meshes - and I don't think any recent implementation should 'take over' a well established term. Agree with 'upstream/downstream' - client and server are usually sufficient. 

My suggestion is to start with the survey of existing configs - Apache httpd, Nginx, servlets, vendor specific - and use the terms and keywords that are most common among them. There is no reason for Gateway API to invent new words or terms, and easier for users to have terms they are already using.

Author: I agree with you costing, if you peek into the prior art section, most implementations call this mTLS, but since this is controversial, I've tried to stick with RFC terms as much as possible

## Non-Goals
- Define other fields that can be used to verify the client certificate such as the Cerificate Hash or Subject Alt Name. 

### API

* Introduce a `clientValidation` field with [Gateway.Tls][] that can be used to validate the client intiating the TLS connection
to the Gateway
* Introduce a `caCerficateRefs` field within `clientValidation` that can be used to specify a list of CA Certificates that
can be used as a trusted anchor to validate the certificates presented by the client

#### GO

```go
// ClientValidationContext holds configuration that can be used to validate the client intiating the TLS connection
// to the Gateway.
// By default, no client specific configuration is validated.
type ClientValidationContext struct {
    // CACertificateRefs contains one or more references to
    // Kubernetes objects that contain TLS certificates of
    // the Certificate Authorities that can be used to
    // as a trusted anchor to validate the certificates presented by the client.
    //

> 	// A single CACertificateRef to a Kubernetes ConfigMap with a key called `ca.crt`

Reviewer 5: Is there an indentation issue?

Reviewer 3: Yeah, looks like the old tab/space indentation dilemma.

Author: thanks for catching this, fixed !

## Non-Goals

> - Define other fields that can be used to verify the client certificate such as the Cerificate Hash or Subject Alt Name. 

Reviewer: Can you point out which of the TLS (or mTLS) use cases in https://docs.google.com/document/d/17sctu2uMJtHmJTGtBi_awGB0YzoCLodtR6rUNmKMCs8/edit#heading=h.cxuq8vo8pcxm that are in your goal?

Reviewer 2: Use case number 7, I believe. Are you just asking or suggesting the GEP should include a link?

Author: as frank mentioned this is 7, does the doc have to be linked into this GEP ?

Reviewer: @arkodg please add a reference to the use case name that's in your goal ( and the doc ).  I strongly believe that tying GEPs in to existing work helps us realize what we've accomplished and what we're still missing from that original list.

Author: done

> ### API

Reviewer: I'd like to know for which route types (HTTP, TLS, TCP, UDP, GRPC, etc) , protocols, and termination types (edge, passthrough?) this might be applicable?

Author: this is same as the downstream info mentioned in https://gateway-api.sigs.k8s.io/guides/tls/#clientserver-and-tls, does this need to be added in the GEP ?

> * Introduce a `clientValidation` field with [Gateway.Tls][] that can be used to validate the client intiating the TLS connection

Reviewer: ```suggestion
* Introduce a `clientValidation` field within [GatewayTlsConfig][] that can be used to validate the client initiating the TLS connection
```

to the Gateway

> * Introduce a `caCerficateRefs` field within `clientValidation` that can be used to specify a list of CA Certificates that

Reviewer: ```suggestion
* Introduce a `caCertificateRefs` field within `clientValidation` that can be used to specify a list of CA Certificates that
```

> can be used as a trusted anchor to validate the certificates presented by the client

Reviewer: ```suggestion
can be used as a trust anchor to validate the certificates presented by the client
```

> 

Reviewer: `GatewayTLSConfig` already has a `CertificateRefs` field.  Why not just add the `CACertificateRefs` field there?https://github.com/kubernetes-sigs/gateway-api/blob/800024e13f634545934ae10d9de82513f6f2798c/apis/v1beta1/gateway_types.go#L374

Author: because there are multiple attributes that can be used to validate the client, and this GEP starts off with CACertRef
here is an exhaustive list supported in Envoy https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#envoy-v3-api-msg-extensions-transport-sockets-tls-v3-certificatevalidationcontext

Reviewer: In that case, I don't quite understand how this GEP ends up so brief compared to the GEP for BackendTLSPolicy.  There seems to be so much missing in terms of what you'd you need for working client certificate validation.

It would probably make sense to add at least:
- why we need this
- a description of use cases for mesh and non-mesh
- a section on how the request flow changes, using the GAMMA request flow in https://gateway-api.sigs.k8s.io/concepts/gamma/?h=request+flow#request-flow and/or the original request flow in https://gateway-api.sigs.k8s.io/concepts/api-overview/?h=request+flow#request-flow
- something that describes how the client certificate validation interacts with edge termination and backend TLS termination
- a section on prior art, e.g. how this is implemented already in a few implementations.

@youngnick do you agree?

Reviewer 3: @candita yes, I think this GEP would be well served by adding some things:
- why we need this - "It's table stakes and widely supported", worded better than that is probably okay, but it should be recorded in the GEP. And the "widely supported" part would at least need a table listing some proxy implementations and their support. 
- I think this also needs a "Deferred" section that covers, at a minimum, "How do I use the system certs?". I think that the custom CA use case will be far more common, but I could be wrong.
- I agree that this GEP should specify what happens when you don't have other TLS details (it doesn't work or gets rejected), and what happens when BackendTLSPolicy is enabled (the client certificate will be used for the Listener connection only, and _not_ reused for the backend connection). This last item - reusing the Listener client cert on a backend connection - feels like something folks may ask for, but we should definitely have out of scope could be another one for the "Deferred" section.

#### GO

```go

> // ClientValidationContext holds configuration that can be used to validate the client intiating the TLS connection

Reviewer: ```suggestion
// ClientValidationContext holds configuration information that can be used to validate the client initiating the TLS connection
```

// to the Gateway.
// By default, no client specific configuration is validated.

> type ClientValidationContext struct {

Reviewer 6: I think here the `ClientValidationContext` should claim some unmentioned behaviors:

1. What if the `caCertificateRefs` field is empty? It's implementation-specific, or all the clients will be rejected, or passed?
2. What's the verify depth of the client certificate? Only the server certificate and the intermediate CA, or the whole chain (Maybe we can add another field like `VerifyDepth`)?

Reviewer 3: Great questions - my thoughts:

1. We could copy the pattern we used for BackendTLSPolicy here, and have something that suggests whether or not to use the system CAs.
2. I think that unless we have a _very_ good reason, we should start with mandating the behavior that the full chain is checked, and then we can add configurability later if we need it. We should make the default case as secure as possible, and not have easily-accessible knobs that make things less secure unless we're _sure_ we need them.

Reviewer 6: @youngnick Fair enough, but we should clarify them in the GEP to reduce the ambiguity

Author: I've add a doc string, mentioning that this is implementation specific today, because im seeing different default verifyDepth settings in different controller

    // CACertificateRefs contains one or more references to
    // Kubernetes objects that contain TLS certificates of

>     // the Certificate Authorities that can be used to
>     // as a trusted anchor to validate the certificates presented by the client.

Reviewer: ```suggestion
    // the Certificate Authorities that can be used
    // as a trust anchor to validate the certificates presented by the client.
```

    //
    // A single CACertificateRef to a Kubernetes ConfigMap with a key called `ca.crt`
    // has "Core" support.
    // Implementations MAY choose to support attaching multiple certificates to
    // a Listener, but this behavior is implementation-specific.
    //
    // References to a resource in different namespace are invalid UNLESS there
    // is a ReferenceGrant in the target namespace that allows the certificate
    // to be attached. If a ReferenceGrant does not allow this reference, the
    // "ResolvedRefs" condition MUST be set to False for this listener with the

>     // "RefNotPermitted" reason.

Reviewer: Can you also add docs on the behavior when the CACertificateRefs is empty but needed?  Echoing https://github.com/kubernetes-sigs/gateway-api/pull/2273/files#r1334951299 to some extent.

Reviewer: I didn't see an update for this.

Author: im not sure I understand this, `caCaertificateRefs` is an optional field within `clientValidation`, if its empty it is assumed that is the intent and it is valid, because in the future `clientValidation` may have more fields within it such as `san`

    //

>     // +kubebuilder:validation:MaxItems=64

Reviewer: It's easier to change this from smaller to larger if needed.
```suggestion
    // +kubebuilder:validation:MaxItems=8
```

    // +optional
    CACertificateRefs []corev1.ObjectReference `json:”caCertificateRefs,omitempty”`
}

```

#### YAML

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: mtls-basic
spec:
  gatewayClassName: acme-lb
  listeners:
  - name: foo-https
    protocol: HTTPS
    port: 443
    hostname: foo.example.com
    tls:
      certificateRefs:
      - kind: Secret
        group: ""
        name: foo-example-com-cert
      clientValidation:
        caCertificateRefs:
        - kind: ConfigMap
          group: ""

>           name: foo-example-com-ca-cert

Reviewer 2: I think Istio users would like to be able to point this to a Secret (instead of ConfigMap), specifically the same secret that `certificateRefs` point to, only with an additional `ca.crt` key containing the client's CA cert.

@howardjohn WDYT?

Reviewer 7: Yes, secret reference is common practice in CNCF landscape, pioneered by cert-manager. Not in the core `tls` SEcret type, but its de-facto standard.

I don't mind config map as well of course.

Reviewer 8: This issue is still live? 

In istio I can do that with:
```yaml
    tls:
      mode: Terminate
      certificateRefs:
      - name: httpbin-credential
      options:
        gateway.istio.io/tls-terminate-mode: MUTUAL
```

I try to setup gateway api with cilium (1.15.5) withtout istio and other sidecars, but MTLS not working
```yaml
    tls:
      certificateRefs:
      - kind: Secret
        name: demo-cert
      frontendValidation:
        caCertificateRefs:
        - kind: Secret
          group: ""
          name: demo-cert
```
